holmes/core/conversations.py

```diff
@@ -0,0 +1,306 @@
+            {
+                "role": "system",
+                "content": system_prompt_without_tools,
+            },
+            {
+                "role": "user",
+                "content": user_prompt,
+            },
+        ]
+        tool_size = calculate_tool_size(
+            ai, messages_without_tools, number_of_tools_for_investigation
+        )
+
+        truncated_investigation_result_tool_calls = [
+            ToolCallConversationResult(
+                name=tool.name,
+                description=tool.description,
+                output=tool.output[:tool_size],
+            )
+            for tool in tools_for_investigation
+        ]
+
+        truncated_template_context = {
+            "investigation": investigation_analysis,
+            "tools_called_for_investigation": truncated_investigation_result_tool_calls,
+            "issue": issue_chat_request.issue_type,
+        }
+        system_prompt_with_truncated_tools = load_and_render_prompt(
+            template_path, truncated_template_context
+        )
+        return [
+            {
+                "role": "system",
+                "content": system_prompt_with_truncated_tools,
+            },
+            {
+                "role": "user",
+                "content": user_prompt,
+            },
+        ]
+
+    conversation_history.append(
+        {
+            "role": "user",
+            "content": user_prompt,
+        }
+    )
+    number_of_tools = len(tools_for_investigation) + len(
+        [message for message in conversation_history if message.get("role") == "tool"]
+    )
+
+    if number_of_tools == 0:
+        return conversation_history
+
+    conversation_history_without_tools = [
+        message for message in conversation_history if message.get("role") != "tool"
+    ]
+    template_context_without_tools = {
+        "investigation": investigation_analysis,
+        "tools_called_for_investigation": None,
+        "issue": issue_chat_request.issue_type,
+    }
+    system_prompt_without_tools = load_and_render_prompt(
+        template_path, template_context_without_tools
+    )
+    conversation_history_without_tools[0]["content"] = system_prompt_without_tools
+
+    tool_size = calculate_tool_size(
+        ai, conversation_history_without_tools, number_of_tools
+    )
+
+    truncated_investigation_result_tool_calls = [
+        ToolCallConversationResult(
+            name=tool.name, description=tool.description, output=tool.output[:tool_size]
+        )
+        for tool in tools_for_investigation
+    ]
+
+    template_context = {
+        "investigation": investigation_analysis,
+        "tools_called_for_investigation": truncated_investigation_result_tool_calls,
+        "issue": issue_chat_request.issue_type,
+    }
+    system_prompt_with_truncated_tools = load_and_render_prompt(
+        template_path, template_context
+    )
+    conversation_history[0]["content"] = system_prompt_with_truncated_tools
```

Comment:
can you explain this?

Reply:
We always expect conversation_history to be in the format we returned it back, and we expect the first item to be this 
{"role": "system", "content": "rendered_prompt"}

Reply:
ok, thx